service_container/service_locators.rst

```diff
@@ -0,0 +1,155 @@
+.. index::
+    single: DependencyInjection; Service Locators
+
+Service Locators
+================
+
+What is a Service Locator
+-------------------------
+
+Sometimes, a service needs the ability to access other services without being sure
+that all of them will actually be used.
+
+In such cases, you may want the instantiation of these services to be lazy, that is
+not possible using explicit dependency injection since services are not all meant to
+be ``lazy`` (see :doc:`/service_container/lazy_services`).
+
+Service Locators are intended to solve this problem by giving access to a set of
+identified services while instantiating them only when really needed, providing
+laziness.
+
+A real-world example being a CommandBus which maps CommandHandler services by Command
+class names and use them to handle their respective command when it is asked for::
+
+    // ...
+    class CommandBus
+    {
+        /**
+         * @var CommandHandler[]
+         */
+        private $handlerMap;
+
+        // ...
+
+        public function handle(Command $command)
+        {
+            if (isset($handlerMap[$commandClass = get_class($command)])) {
+                $handler = $handlerMap[$commandClass];
+
+                return $handler->handle($command);
+            }
+        }
+    }
+
+    // ...
+    $commandBus->handle(new FooCommand());
+
+Because only one command is handled at a time, other CommandHandler services are not
+used.
```

Comment:
done, also moved the "Service locators are intended to solve this problem" just after this "wrong" example in order to know what it solves before reading it